src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -1340,31 +1345,43 @@ std::optional<segment_meta> async_manifest_view::search_spillover_manifests(
           return -1;
       },
       [&](model::timestamp t) {
+          if (manifests.empty()) {
+              return -1;
+          }
           vlog(
             _ctxlog.debug,
             "search_spillover_manifest query: {}, num manifests: {}, first: "
             "{}, last: {}",
             query,
             manifests.size(),
-            manifests.empty() ? model::timestamp{}
-                              : manifests.begin()->base_timestamp,
-            manifests.empty() ? model::timestamp{}
-                              : manifests.last_segment()->max_timestamp);
-          const auto& bt_col = manifests.get_base_timestamp_column();
-          const auto& mt_col = manifests.get_max_timestamp_column();
-          auto mt_it = mt_col.lower_bound(t.value());
-          if (mt_it.is_end()) {
+            manifests.begin()->base_timestamp,
+            manifests.last_segment()->max_timestamp);
+
+          auto first_manifest = manifests.begin();
+          auto base_t = first_manifest->base_timestamp;
+          auto max_t = manifests.last_segment()->max_timestamp;
+
+          // Edge cases
+          if (t < base_t || max_t == base_t) {
+              return 0;
+          } else if (t > max_t) {
               return -1;
           }
-          auto bt_it = bt_col.at_index(mt_it.index());
+
+          const auto& bt_col = manifests.get_base_timestamp_column();
+          const auto& mt_col = manifests.get_max_timestamp_column();
+          auto mt_it = mt_col.begin();
+          auto bt_it = bt_col.begin();
+          int target_ix = -1;
           while (!bt_it.is_end()) {
-              if (t.value() >= *bt_it && t.value() <= *mt_it) {
-                  return static_cast<int>(bt_it.index());
+              target_ix = static_cast<int>(bt_it.index());
+              if (*mt_it >= t.value() || *bt_it > t.value()) {
```

Comment:
when you flipped the `bt_it`/`t.value()` inequality it went from `<=` to `>` as expected. but the `mt_it`/`t.value()` inequality kept the `=` condition. was that intentional to have the logic change?

Reply:
Presumably you're talking about this: https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/partition_manifest.cc#L2412-L2422

Given the hard relationship we're trying to impose, could we move this into some `find_index_for_timestamp` template and reuse it here?